include/envoy/stats/stats.h

```diff
@@ -408,25 +408,36 @@ typedef std::unique_ptr<StoreRoot> StoreRootPtr;
 struct RawStatData;
 
 /**
- * Abstract interface for allocating a RawStatData.
+ * Abstract interface for allocating statistics. Implementations can
+ * be created utilizing a single fixed-size block suitable for
+ * shared-memory, or in the heap, allowing for pointers and sharing of
+ * substrings, with an opportunity for reduced memory consumption.
  */
-class RawStatDataAllocator {
+class StatDataAllocator {
 public:
-  virtual ~RawStatDataAllocator() {}
+  virtual ~StatDataAllocator() {}
 
   /**
-   * @return RawStatData* a raw stat data block for a given stat name or nullptr if there is no
-   *         more memory available for stats. The allocator should return a reference counted
-   *         data location by name if one already exists with the same name. This is used for
-   *         intra-process scope swapping as well as inter-process hot restart.
+   * @param name the full name of the stat.
+   * @param tag_extracted_name the name of the stat with tag-values stripped out.
+   * @param tags the extracted tag values.
+   * @return Counter* a counter, or nullptr if allocation failed.
    */
-  virtual RawStatData* alloc(const std::string& name) PURE;
+  virtual CounterSharedPtr makeCounter(const std::string& name, std::string& tag_extracted_name,
```

Comment:
`tag_extracted_name` and `tags` are passed as references not so they can be modified and read by the caller, but so that they can be moved.  I think a better way to communicate this in the interface is by making them rvalue references (for example: `std::string&& tag_extracted_name`), which will force all callers to wrap them in `std::move()` or pass an actual rvalue.

Also, the interface should document that if the returned CounterSharedPtr is `nullptr`, then these objects will not be moved/modified (we assume this at the callsite).

Similar comments for `makeGauge()`.

Reply:
done